### PR TITLE
initial draft for evaluating Data Channel as the transport protocol.

### PR DESCRIPTION
**Pull request by schien**
address issue #10

### CODE REVIEW

File: datachannel.md [hunk 1]
```diff
@@ -0,0 +1,171 @@
+# Data Channel
+
+[Data Channel](https://tools.ietf.org/html/draft-ietf-rtcweb-data-channel-13)
+is a non-media transport protocol in WebRTC framework, designed for exchanging
+data from peer to peer. It advertise the following benefits:
```
Comment (mfoltzgoogle) [inline datachannel.md]: advertises

File: datachannel.md [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+
+- Implemented in major browsers (except Safari)
+- Support heterogeneous network environment e.g. NAT, firewall.
```
Comment (mfoltzgoogle) [inline datachannel.md]: Supports

File: datachannel.md [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+- Easily support media transport for remote playback
```
Comment (mfoltzgoogle) [inline datachannel.md]: Supports

File: datachannel.md [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+# Design and Specification
+
+- [WebRTC overview](https://tools.ietf.org/html/draft-ietf-rtcweb-overview-18)
+- IETF draft: [WebRTC Data Channel Establishment Protocol](https://tools.ietf.org/html/draft-ietf-rtcweb-data-protocol-09)
+- IETF draft: [Data Channel](https://tools.ietf.org/html/draft-ietf-rtcweb-data-channel-13)
+
+# Presentation API functionality
+
+Data Channel is designed to be a non-media transport for WebRTC, with built-in
+sub protocol negociation mechanism defined.
```
Comment (mfoltzgoogle) [inline datachannel.md]: Typo in "negotiation"

File: datachannel.md [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+
+A Data Channel is built on top of SCTP/DTLS/UDP protocol stack, which supports
+reliable data transmition. Each Data Cahnnel has a unique integer ID and a label.
```
Comment (mfoltzgoogle) [inline datachannel.md]: Typo in Channel

File: datachannel.md [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+Mutiple Data Channels can share the same DTLS connection.
+
+Assume that a single Data Channel is used as a control channel to transmit control
+messages between the controller and receiver and one Data Channel is created for each
+connected Presentation Connection for trasmitting application data.
```
Comment (mfoltzgoogle) [inline datachannel.md]: Typo: transmitting

File: datachannel.md [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+
+An extra signalling channel and protocol for exchanging ICE candidates is required
+to bootstrap the control channel. The bootstraping procedure needs to be executed
```
Comment (mfoltzgoogle) [inline datachannel.md]: bootstrapping

File: datachannel.md [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+2. Receiver spawn a Data Channel on the same RTC connection of control channel, with
+   presentation ID as label.
+3. Get a reply on control channel with the ID and URL to confirm success
+   (or report an error).
+
+## Presentation Resumption
+
+1. Send a control message on control channel from the controller to the display
+   reconnect to a presentation with ID and URL for the initial presentation connection.
+2. Receiver spawn a Data Channel on the same RTC connection of control channel, with
```
Comment (mfoltzgoogle) [inline datachannel.md]: If a second controller wants to initiate connection to the same presentation, would it have a duplicate Data Channel label?

File: datachannel.md [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+# Reliability
+
+WebRTC is designed to be operated on heterogeneous network environment, using
+ICE for address discovery and built-in negociation/re-negociation mechanism.
+
+*TODO:* Get reliability data.
+
+# Latency to establish a new connection
+
+Data on connection establishment latency.
```
Comment (mfoltzgoogle) [inline datachannel.md]: It sounds like are three sources of latency:
1. Connecting a reliable channel the first time for for ICE negotiation.
2. Establishing the first Data Channel (perhaps this is part of #1).
3. Adding an additional Data Channel on top of an existing peer-to-peer connection.

File: datachannel.md [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+
+WebRTC framework supports media codec negociation and real-time media transport.
+
+**TODO:** Fill in when Remote Playback equirements are known.
+See [Issue #3](https://github.com/webscreens/openscreenprotocol/issues/3).
+
+# Reliability
+
+WebRTC is designed to be operated on heterogeneous network environment, using
+ICE for address discovery and built-in negociation/re-negociation mechanism.
```
Comment (mfoltzgoogle) [inline datachannel.md]: s/negociation/negotiation/g

**Comment by mfoltzgoogle** [comment]
This looks great @schien!  Most of my comments are minor typos.

Did you have thoughts about establishing the bootstrap reliable channel for ICE negotation?  IIRC Mozilla uses a TCP channel for this.  I believe the bootstrap channel needs to remain open in case the network routing changes and ICE renegotiation occurs.

I recently heard from local WebRTC contacts that there is exploratory work to layer RTCDataChannel on top of QUIC, since the functionality is similar.  Something to keep an eye on.

**Comment by schien** [comment]
Yes, the example protocol we have in current Gecko codebase uses TCP channel for bootstrapping DataChannel. To be more specific, TLS connection is used (with self-generated certificate, fingerprint broadcast on DNSSD).
We might use QUIC to replace the TLS/TCP channel if maintaining a QUIC connection is cheaper.

**Comment by anssiko** [comment]
@schien, great work! Feel free to land this. You can keep #10 open to track the remaining TODOs.

File: datachannel.md (hunk 8 above)
Comment (schien) [inline datachannel.md]: In order to support many-to-one presentation, additional controller Id is required for pairing the correct data structure in receiver side.

File: datachannel.md (hunk 9 above)
Comment (schien) [inline datachannel.md]: Yes, exactly. I'll add it to this document.

**Pull request merged**